### ISSUE TITLE
CA1859 (use concrete types) suggestion breaks compilation for interface

### ISSUE THREAD

**Issue by mikeclayton** [dotnet/roslyn#83152]
This seems a bit too obvious to be a bug, so apologies in advance if it's just noise, but it's generating a warning in the IDE and it would be good if it didn't...

---

[CA1859 - Use concrete types when possible for improved performance](https://learn.microsoft.com/en-gb/dotnet/fundamentals/code-analysis/quality-rules/ca1859) is being flagged on code where changing to the concrete implementation would break compilation.

For example, following the recommendation in the CS1856 warning in the code below and changing the return type of ```CreateMyInterface``` to ```MyClass``` causes a CS1061 error further down the code:

```csharp
namespace MyNamespace;

public interface IMyInterface
{
    public void DoSomething();
}

public sealed class MyClass : IMyInterface
{
    public void DoSomething()
    {
    }
}

public static class MyStaticClass
{
    private static IMyInterface CreateMyInterface() => new MyClass();
    //                          ^^^^^^^^^^^^^^^^^
    //                          CS1859 - Change return type of method 'CreateMyInterface'
    //                          from 'MyNamespace.IMyInterface' to 'MyNamespace.MyClass'
    //                          for improved performance

    public static void Main()
    {
        var myClass = CreateMyInterface();
        myClass.DoSomething();
        //      ^^^^^^^^^^^
        //      CS1061 - 'MyClass' does not contain a definition for 'DoSomething'
        //      and no accessible extension method 'DoSomething' accepting
        //      a first argument of type 'MyClass' could be found (are you
        //      missing a using directive or an assembly reference?)
    }
}
```

<img width="1244" height="604" alt="Image" src="https://github.com/user-attachments/assets/326e0856-c75d-4c48-92d3-8e29dd1e171c" />

It's not reproducible on sharplab, but here's a link anyway:

[CA1859 - Use concrete types when possible for improved performance](https://sharplab.io/#v2:CYLg1APgdghgtgUwM4AcYGMEAICyBPAOXmTUwG4BYAKGoAEBmLASygBcEAnAMw2wEl8fNpx6ZqAb2pZpWBrIAsWACIB7AMorErABYsA5gAoAlJSoBfanUZIEMADYJgsgEy48AYTswkSLCCwCeELs3LwSUjJytIqqGlq6UIZGEdKSVDJYFuaWVFEAjABsLm5qrDCsTOie3kjh6TIoHEwAbuXYtIUBgsKhmFjuHLbs3SGiCMZYALwAfFhQCADubtU+xqYZAPQbGTu7exkAekfHJ6dHKVhb+9fX7mp5ABwArACcWAC0/dowie0A7FhWHgUNgVFwsPEVE4AOQDIYIEYiXjQi5XG7ojJcDiaLDQ/BERCoXgAOkCwSRmGhgJUuPxxCJmGJ+BWSBR9WkaIxGK4Kg4zDgjRUzUcWBB3F5cB+Yho7JczguaX2rT5cA8Xh8U36gzaiN64xMFwyqpZxNimgQOn0a0NMk5MjOpxtHO2OzueQADAU8h9aWqalTgCpkHMVKwsOgVGwYCwsDAsMAEFwWEwKpGsDy+dCzfF9Gz9nbpD8nFAaRhMD4mAAjBxYBAAD3YUCQTDTkJh2YtCT0VLLCBQFUSTsuLoycaTHCQYZgHD0AFdEGwsGDAcDsHi/T4qRHZ3YnJXsDzZ1AnAZp9g8CpZ0OCxCmBXErHZO740xBugKsKl3yfrGfAg4NWeBYIMXCcAgUCYAA/MkspZGYQA=)

**Version Used**: 

* Visual Studio 2026 Insiders [11626.173]

```
C:\> dotnet --info
.NET SDK:
 Version:           10.0.201
 Commit:            4d3023de60
 Workload version:  10.0.200-manifests.34a88a22
 MSBuild version:   18.3.0-release-26153-122+4d3023de6

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.26100
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\10.0.201\
```

**To reproduce**

* Paste the above code into a new project
* Notice the CS1859 error
* Replace this line to resolve CS1859:
  * from ```private static IMyInterface CreateMyInterface() => new MyClass();```
  * to ```private static MyClass CreateMyInterface() => new MyClass();```
* CS1061 is flagged instead

**Diagnostic Id**:

CA1859 - Use concrete types when possible for improved performance

**Expected Behavior**:

The analyzer should not suggest using a concrete type where changing it would cause compilation errors.

**Actual Behavior**:

The analyzer suggests a using concrete type where changing it would cause compilation errors.

**Comment by CyrusNajmabadi** [comment]
The goal of this isn't to keep the semantics the same.  It's to ensure that you are passing back as an interface intentionally.   If so, you don't take the suggestion there. It's more information rather than something you are supposed to follow in all cases.

**Comment by mikeclayton** [comment]
Cool. Thanks for the quick reply.

I can suppress the warning so it's not a big deal. I just assumed analysers would only recommend things that didn't break compilation so I was a bit surprised this one came up and figured I'd check.

I'll close this now...
